arvo/eyre/external-api-ref.md

```diff
@@ -0,0 +1,399 @@
+
+The `cookie` header contains the session cookie obtained by [authenticating](#authentication). The data is a JSON array containing one or more JSON [action](#actions) objects. The channel `uid` is a unique name of your choosing. Typically you'd use the current unix time and a hash but there's no specific requirements. If you're opening a new channel you'd choose a new `uid` and if you already have a channel open you'd just use that.
+
+If successful, Eyre will respond with a status code of 204. Note the HTTP response will have no content, any [responses](#responses) will be sent as SSEs on the channel event stream.
+
+## Actions
+
+### Poke
+
+This is for poking a Gall agent.
+
+| Key      | JSON Type | Example Value | Description                                                   |
+|----------|-----------|---------------|---------------------------------------------------------------|
+| `id`     | Number    | `1`           | Sequential ID for keeping track of sent messages.             |
+| `action` | String    | `'poke'`      | The kind of action.                                           |
+| `ship`   | String    | `'zod'`       | Ship name, excluding leading `~`.                             |
+| `app`    | String    | `'hood'`      | Name of the gall agent you're poking.                         |
+| `mark`   | String    | `'helm-hi'`   | Type of data. Must correspond to a mark definition in `/mar`. |
+| `json`   | Any       | `'hello'`     | Actual payload. Any JSON type, determined by app.             |
+
+**Example**
+
+```json
+{
+  'id': 1,
+  'action': 'poke',
+  'ship': 'zod',
+  'app': 'hood',
+  'mark': 'helm-hi',
+  'json': 'hello'
+}
+```
+
+**Response**
+
+Eyre will send a [poke ack](#poke-ack) as an SSE event on the channel event stream.
+
+### Subscribe
+
+This is for subscribing to a watch `path` of a Gall agent.
+
+| Key      | JSON Type | Example Value   | Description                                         |
+|----------|-----------|-----------------|-----------------------------------------------------|
+| `id`     | Number    | `2`             | Sequential ID for keeping track of sent messages.   |
+| `action` | String    | `'subscribe'`   | The kind of action.                                 |
+| `ship`   | String    | `'zod'`         | Ship name, excluding leading `~`.                   |
+| `app`    | String    | `'graph-store'` | Name of the gall agent to which you're subscribing. |
+| `path`   | String    | `'/updates'`    | The path to watch. Depends on the app.              |
+
+**Example**
+
+```json
+{
+  'id': 2,
+  'action': 'subscribe',
+  'ship': 'zod',
+  'app': 'graph-store',
+  'path': '/updates'
+}
+```
+
+**Response**
+
+Eyre will send back a [watch ack](#watch-ack). If subscribing was successful, you will also begin receiving any [fact](#fact)s sent by the Gall agent on the specified `path`.
+
+### Ack
+
+This is for acknowledging a [fact](#fact) from a Gall agent to which you've subscribed.
```

Comment:
the only other possible responses are a kick, poke-ack and watch-ack, and afaict none of those require an ack, only a `diff` type response which is only triggered by a fact (and they can only come from agents cos you can't poke or watch vanes)

```
          %fact
        :~  ['response' [%s 'diff']]
          ::
            :-  'json'
            ~|  [%unexpected-fact-mark p.cage.sign]
            ?>  =(%json p.cage.sign)
            ;;(json q.q.cage.sign)
        ==
```

Reply:
hmm no perhaps you're right, it seems it wants them for all. I was confused I think cos it's weird to ack an ack

Reply:
You aren't acking the `poke-ack` itself, you are acking the SSE message that notified you of the `poke-ack`.

This is more obvious in the broader context: Eyre keeps a queue of message it has sent over the channel, because it doesn't know whether you're still connected or not. (And might want to send you the ones you missed when you reconnect.) Acking a message lets Eyre remove it from the queue, because you've confirmed you did indeed receive it.